cli_tools/gce_vm_image_import/importer/source.go

```diff
@@ -0,0 +1,198 @@
+	sourceBucketName, sourceObjectName, err := storage.GetGCSObjectPathElements(gcsPath)
+	if err != nil {
+		return nil, err
+	}
+	source := fileSource{
+		gcsPath: gcsPath,
+		bucket:  sourceBucketName,
+		object:  sourceObjectName,
+	}
+	return source, source.validate(storageClient)
+}
+
+// The resource path for fileSource is its GCS path.
+func (s fileSource) path() string {
+	return s.gcsPath
+}
+
+// Performs basic validation, focusing on error cases that we've seen in the past.
+// This reads a few bytes from the file in GCS. It is an error if the file
+// is empty, or if the file is compressed with gzip.
+func (s fileSource) validate(storageClient domain.StorageClientInterface) error {
+	rc, err := storageClient.GetObjectReader(s.bucket, s.object)
+	if err != nil {
+		return daisy.Errf("failed to read GCS file when validating resource file: unable to open "+
+			"file from bucket %q, file %q: %v", s.bucket, s.object, err)
+	}
+	defer rc.Close()
+
+	byteCountingReader := daisycommon.NewByteCountingReader(rc)
+	// Detect whether it's a compressed file by extracting compressed file header
+	if _, err = gzip.NewReader(byteCountingReader); err == nil {
+		return daisy.Errf("the input file is a gzip file, which is not supported by " +
+			"image import. To import a file that was exported from Google Compute " +
+			"Engine, please use image create. To import a file that was exported " +
+			"from a different system, decompress it and run image import on the " +
+			"disk image file directly")
+	}
+
+	// By calling gzip.NewReader above, a few bytes were read from the Reader in
+	// an attempt to decode the compression header. If the Reader represents
+	// an empty file, then BytesRead will be zero.
+	if byteCountingReader.BytesRead <= 0 {
+		return daisy.Errf("cannot import an image from an empty file")
+	}
+
+	return nil
+}
+
+// An importable source backed by a GCE disk image.
+type imageSource struct {
+	uri string
+}
+
+// Creates an imageSource from a reference to a GCE disk image. The syntax of the
+// reference is validated, but no I/O is performed to determine whether the image
+// exists or whether the calling user has access to it.
+func newImageSource(imagePath string) (resource, error) {
+	source := imageSource{
+		uri: param.GetGlobalResourcePath("images", imagePath),
+	}
+	return source, source.validate()
+}
+
+var imageNamePattern = regexp.MustCompile("^[a-z]([-a-z0-9]*[a-z0-9])?$")
+
+// Performs basic validation, focusing on error cases that
+// we've seen in the past.
+//
+// Two key offenders:
+//   1. Using a file path instead of an image name.
+//     Examples:
+//        - file://home/user/image.vmdk
+//        - gs://bucket/image.vmdk
+//        - gs://global/images/image
+//        - https://storage.googleapis.com/bucket/image
+//   2. Using a relative file path instead of an image name.
+//     Examples:
+//        - image.vmdk
+//        - path/to/image.vmdk
```

Comment:
I feel the "offenders" would be better documented in tests instead of in method docs (it would be self documenting). This is really verbose and reduces readability IMO. It will also probably become outdated really fast once we find new offenders. That's why tests are a better documentation since they will have to be updated while docs are easily forgotten.